src/services/PeerDatabase.ts

```diff
@@ -31,28 +32,67 @@ export default class PeerDatabase {
     return choices(Array.from(this.peers), numPeers)
   }
 
+  public async getFreePeers () {
+    const peersFromDB = await this.codiusdb.getPeers()
+    const freePeers = peersFromDB.filter(peer => {
+      const peerFreeMem = this.memoryMap.get(peer)
+      if (peerFreeMem) {
+        return peer
+      }
+    })
+    return freePeers
+  }
   public async addPeers (peers: string[]) {
     const previousCount = this.peers.size
     for (const peer of peers) {
       if (peer === this.identity.getUri()) {
         continue
       }
-      const memory = await axios.get(peer + '/memory')
-      this.memoryMap.set(peer, memory.data.freeMem)
+      try {
+        // Check for invalid peer addresses. Validate peer only if not in set.
+        if (!this.peers.has(peer)) {
+          const peerInfo = await axios.get(peer + '/host-info')
+          if (peer === peerInfo.data.uri && shallowValidatePeer(peer)) {
+            this.memoryMap.set(peer, peerInfo.data.fullMem)
+            this.peers.add(peer)
+          }
+        }
+
+      } catch (e) {
+        if (process.env.NODE_ENV !== 'test') {
+	  log.error('Error code %s at %s', e, peer)
+        }
+      }
 
-      this.peers.add(peer)
     }
-    if (this.peers.size > previousCount) {
+    if (this.peers.size > previousCount && process.env.NODE_ENV !== 'test') {
       this.codiusdb.savePeers([...this.peers]).catch(err => log.error(err))
       log.debug('added %s peers, now %s known peers', this.peers.size - previousCount, this.peers.size)
     }
   }
+  public removePeer (peer: string) {
+    this.peers.delete(peer)
+    this.codiusdb.savePeers([...this.peers]).catch(err => log.error(err))
+    log.debug('removed peer %s, now %s peers', peer, this.peers.size)
+  }
+
+  public getNumPeers () {
+    return this.peers.size
+  }
+
+  public getAllPeers () {
+    return [...this.peers]
+  }
 
   private async loadPeersFromDB () {
```

Comment:
is this well named, it seems to just do the shallow validation rather than loading the peers for use.